pygfx/utils/show.py

```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
 
         background = Background(None, BackgroundMaterial((0, 1, 0, 1), (0, 1, 1, 1)))
         scene.add(background)
+        scene.add(AmbientLight(0.2))
 
-    camera = PerspectiveCamera(70, 16 / 9)
-    look_at = camera.show_object(object)
-    scene.add(camera)
+    if not camera:
+        camera = PerspectiveCamera(70, 16 / 9)
+        camera.add(DirectionalLight(0.8))
+        scene.add(camera)
 
-    # Add lights if there are none
-    light_count = 0
+    if canvas is None:
+        from wgpu.gui.auto import WgpuCanvas, run
 
-    def check_light(ob):
-        nonlocal light_count
-        if isinstance(ob, Light):
-            light_count += 1
+        canvas = WgpuCanvas()
+        event_loop = run
+    elif event_loop is None:
+        # TODO: find the matching run function
+        raise ValueError(
+            "When providing a canvas you also need to provide the event loop."
+        )
 
-    scene.traverse(check_light, False)
-    if not light_count:
-        camera.add(DirectionalLight(0.8))
-        scene.add(AmbientLight(0.2))
+    if renderer is None:
+        renderer = WgpuRenderer(canvas)
+
+    if controller is None:
+        look_at = camera.show_object(object)
+        controller = OrbitController(camera.position.clone(), look_at, up=up)
+        controller.add_default_event_handlers(renderer, camera)
+
+    if draw_function is None:
+
+        def animate():
+            if before_render is not None:
+                before_render()
```

Comment:
If we'd provided `renderer` as an argument to `before_render` and `after_render` we could e.g. request new draws in the user code.

Reply:
Would that be an advantage? The current default loop already requests a new draw at the end in any case, so I am inclined to say this would be the realm of using a custom `event_loop`, which could totally have a reference to the canvas and the renderer.

Reply:
I would recommend to see if the need comes up while you are rewriting more examples later on.